Source/JavaScriptCore/jit/AssemblyHelpers.h

```diff
@@ -144,38 +144,68 @@ class AssemblyHelpers : public MacroAssembler {
 #endif
     }
 
+#if USE(JSVALUE64)
     template<typename T, typename U>
     void storeCell(T cell, U address)
     {
-#if USE(JSVALUE64)
         store64(cell, address);
+    }
 #else
+    void storeCell(GPRReg cell, Address address)
+    {
         static_assert(!PayloadOffset && TagOffset == 4, "Assumes little-endian system");
-        storePair32(cell, TrustedImm32(JSValue::CellTag), address);
-#endif
+        store32(TrustedImm32(JSValue::InvalidTag), address.withOffset(TagOffset));
+        store32(cell, address);
+        store32(TrustedImm32(JSValue::CellTag), address.withOffset(TagOffset));
```

Comment:
So, what guarantees the ordering of these three stores? ARM v7 deploys weak memory ordering and stores can be reordered freely. This means that the other core can see the memory, only cell is written. (InvalidTag and CellTag stores are reordered and deferred after the cell store)

Reply:
The ISA does require the stores to occur in program order as far as the core performing them is concerned; we're relying on the hardware's cache coherency to do the rest; as long as the stores reach the L1 cache in program order (and, IIUC they must: there's only one L1 cache line involved, reordering them in general might violate the ISA requirement), ... then we'll observe them that way on other cores. 

That is to say, it's hardware-specific and, you're right, not guaranteed by the ISA.

Admittedly, I just double-checked and some older ARMv7 chips (e.g. Cortex-A8) probably _don't_ provide the cache coherency guarantees we're using here; but we only care about Cortex-A15 and some armv8 chips (A53, A55) which do... I think we'll probably need a build flag to control whether or not this code is used.

We've tested the behavior of the store ordering on hardware: at least on a Cortex-A72 and also on our buildbox (which has a Neoverse N-1 ....) it behaves as expected. I have a board with a Cortex-A8 I can test on; but, based on the reference from ARM I'm not sure if it will work--though, because we don't really have multiple cores writing, we might luck out--it's hard for me to say without testing... I'll update here.

Reply:
I think it should be checked that the hardware model of CPU and disable concurrent JIT when it is ARMv7 and unsupported cores in Options.cpp.